### ISSUE TITLE
Using a more posix-compliant way of Popen

### ISSUE THREAD

**Issue by davidhalter** [xonsh/xonsh#3019]
This is a more general issue that I want to discuss with you guys. I really like the idea of xonsh. Awesome job. I wanted to do something similar with parso[1], but then realized that it already exists :)

This is related to #2353, but I'm intentionally opening a new ticket. I want to continue the discussion about threads that was started in https://github.com/xonsh/xonsh/pull/1758#issuecomment-251032267.
I'm not sure if what I'm triggering (see at the bottom) is happening, because it avoids the commands cache, but I'd rather not search for `tcsetattr` in a binary in the first place (if that's what I think it does).

My general impression is that threading makes xonsh's job harder than it should be. The issue below is something that e.g. doesn't happen in a very simplified shell below that I've build over a week. This shell is IMO how it should be: All foreground Popen processes should block. All other processes can be captured or not, it doesn't really matter. IMO there's no need for threading. If there is a need for threading, it's probably Python code. It's not that easy to suspend that. In that case I think threading would make sense. But it's probably possible to still run foreground processes in the main thread and block.

From https://github.com/xonsh/xonsh/pull/1758#issuecomment-251032267:
> The contradiction is that the queue can be empty, the pipe can still have valid data, and the process can be complete.
> Thus, before/while you do the last few reads, you have to wait. But you can't wait indefinitely because that blocks, the process is already over (so you can't wait on that), and you can't wait on the pipe (because it is the thing you are reading from).
> The strategy here is the wait for 55 times the timeout for no new data to come into the queue. If any data does come in ever, the timer reset. As heuristics go, I think it is a pretty effective one. It also has the benefit of being the only way forward I saw that didn't end in actually intractable contradictions.

Is there a way how we can reproduce that? I also didn't find the 55 somewhere. I would be really interested in this case. However in any case, I don't think it changes when we need threading. Well... Maybe we need threading here, but then we could at least ditch it in all the other cases (bash and other shells are not using threading either).

I don't want to be rude, the project is really cool. it's just that the non blocking part breaks some things that annoy me :) e.g. pressing the enter key for a few seconds. I'm happy to discuss this here, or if it takes too much time, we can also have a quick call (skype or whatever).

I'm very happy to help with PRs. However I'm not sure I understand the general code base. So I might need some guidance. Also I'm really not sure how we want to continue with threading. I was trying some things in your code base, but I thought I'd first start a discussion.  

BTW: Once we get this working, I'm also happy to improve Jedi support, if you're interested.

```
def clean_shell():
    p = None
    fileno = sys.stderr.buffer.raw.fileno()
    while True:
        if p is not None:
            result = os.tcsetpgrp(fileno, p.pid)
            pid, status = p._try_wait(os.WUNTRACED)
            if p.poll() is not None:
                p = None
            # I don't understand why this is necessary, but I get a lot of
            # SIGTTOU's that make it impossible to call tcsetpgrp (seen with
            # strace -f).
            old_mask = signal.pthread_sigmask(
                signal.SIG_BLOCK,
                {signal.SIGTTOU, signal.SIGTTIN}
            )
            result = os.tcsetpgrp(fileno, os.getpid())
            signal.pthread_sigmask(signal.SIG_UNBLOCK, old_mask)

        while True:
            print('$ ', end='')
            command = input()
            if command == 'fg':
                os.killpg(p.pid, signal.SIGCONT)
                break
            elif command == 'exit':
                exit()
            elif command:
                if p is None:
                    p = Popen(
                        command,
                        shell=True,
                        preexec_fn=lambda: os.setpgrp(),
                    )
                    break
                else:
                    print("there's still an open process, please use fg")
            else:
                print("unknown command")


if __name__ == '__main__':
    clean_shell()
```

[1]: https://github.com/davidhalter/parso

## xonfig

No configuration

## Expected Behavior

ctrl-z should always work for shell commands.

## Steps to Reproduce

Type `sh -c 'vi'` and Enter.
Hit `ctrl-z` in vi. The shell is now frozen and even typing `ctrl-c` doesn't help.

**Comment by scopatz** [comment]
Hey @davidhalter - thanks for opening this! I think you have correctly identified an area that xonsh could certainly use more eyes.  And we welcome and encourage discussion here (not rude at all)!  If you want to have a chat with me personally you can [schedule one it here](https://scopatz-quansight.youcanbook.me).

> I'm not sure if what I'm triggering (see at the bottom) is happening, because it avoids the commands cache, but I'd rather not search for `tcsetattr` in a binary in the first place (if that's what I think it does).

While `tcsetattr` and other tokens are searched for in commands where we know nothing about them, by default, `sh` [uses](https://github.com/xonsh/xonsh/blob/master/xonsh/commands_cache.py#L423) the [predict_shell()](https://github.com/xonsh/xonsh/blob/master/xonsh/commands_cache.py#L326) function. `tcsetattr` is not checked and so perhaps something is wrong with the logic here?

> IMO there's no need for threading.

I'd really love to see an implementation that doesn't rely on threading! I think this maybe possible with some of the asyncio stuff that we can use starting next month.  (Python v3.4 reaches end-of-life in March). I think in the whole subprocess execution system could be rewritten a lot more cleanly using async and could be done in such a way that avoids a lot of the explicit threading that we have right now. 

I also believe that the notion of a "rich, async command pipeline object" could & should be separated out into its own library. In my mind, the requirements for such a package are:

* Relies only on Python 3.5+ standard library and/or vends its dependencies.
* Works natively on all platforms (including Windows)
* Seemlessly inter-operates subprocess commands and Python functions that conform to the proper signature.
* Separates the pipeline specification from its execution (ie you should be able define a pipeline and then execute it later)
* Supports all combinations of interactivity and capturing stdout/stderr

If we had a library that did that, then it would be xonsh-ready, as this is what xonsh does now. @astronouth7303 made an attempt to do this with [`xonsh/slug`](https://github.com/xonsh/slug). We could revive this or pursue an new attempt based on async. 

I am very happy to help work on something that does this, I just haven't wanted to do it alone and most of the other xonsh devs have learned to stay away from the subprocess model :wink:  

> I also didn't find the 55 somewhere

I think this might have increases to 1000x. See https://github.com/xonsh/xonsh/blob/master/xonsh/proc.py#L2467

>  I'm also happy to improve Jedi support, if you're interested.

That's be awesome!

PS parso is really cool! I wonder if we could move to that rather than our current PLY implementation?

**Comment by scopatz** [comment]
> I'm very happy to help with PRs. However I'm not sure I understand the general code base. So I might need some guidance.

Awesome! I am happy to give you a tour too, if you want!

**Comment by davidhalter** [comment]
Thanks for your answer!

I will try to learn a bit about how other shells does pipelining, because I don't really understand what happens with process groups and signals in that case. I will write my findings here. It will take time. It might take me a month or two. It took me a few weeks to learn about process groups, WUNTRACED and all the other things about terminals, so that will also take a while :)

> PS parso is really cool! I wonder if we could move to that rather than our current PLY implementation?

Well, that's what I was thinking about as well. At the moment I don't see the urgency of changing it, because that part off xonsh is working really well IMO. However looking at the code there seem to be quite a few "hacks" to get it working. I think a parser generator would be way nicer. Also ply is just 
"a horrible abuse of docstrings" (quoting David Beazley himself, https://www.youtube.com/watch?v=zJ9z6Ge-vXs). I really like the creation of the AST though. We would definitely also need something like that with parso as well. So the current way is working pretty well, no need for quick changes.

> I also believe that the notion of a "rich, async command pipeline object" could & should be separated out into its own library. In my mind, the requirements for such a package are:

There are some primitives even in 3.4: https://docs.python.org/3.4/library/asyncio-subprocess.html#asyncio-subprocess I'm however not sure if that helps, because I really think we should be using `os.WUNTRACED`. So we might need to rewrite some things. I'm also still not 100% sure if any form of parallelism except maybe threading for Python code is needed. But I still don't know enough about the linux kernel. For Windows I'm even less sure.

**Comment by scopatz** [comment]
Well, anytime you have questions or want to chat about it let me know!
